### ISSUE TITLE
TextEncoder format is difficult to parse

### ISSUE THREAD

**Issue by rogpeppe** [uber-go/zap#189]
The TextEncoder doesn't quote field values, which means that
if fields contain spaces or equal characters, it's not possible in general to
say where the field boundaries are.

Also, because it just uses fmt.Sprintf("%+v") to format objects,
information will be lost if the objects contain pointers or have custom
JSON marshaling methods.

The right answer might be to encode to JSON anyway, but post-process
into a more human-readable format - it would be nice if the
TextEncoder output contained all the info that would be in the usual
JSONEncoder output.

**Comment by akshayjshah** [comment]
Agreed - the current `TextEncoder` is still pretty rough. I'm a bit leery of using `MarshalJSON` if the output isn't actually JSON, but we should certainly use `MarshalText` where available.

What are your thoughts on multiline logging output? Internally, we only use the text output for human consumption (usually tee'd to standard out in development and testing), so something YAML-ish is appealing.
```
[I] 1970-01-01T12:00:00Z This is an info message.
     "foo": "bar"
     "baz": {"complex": "object"}
     "someList": ["a", "b", "c"]
```

This is rather difficult to machine-parse, but is nice for humans. It also makes it easier to format stack traces and other complex fields.

**Comment by rogpeppe** [comment]
> I'm a bit leery of using MarshalJSON if the output isn't actually JSON

I know what you mean, but I think it might be for the best, otherwise
you are never going to get information parity between the different
encoder outputs, because many types will implement MarshalJSON
but not MarshalText, and vice versa, and I think that having the
text output represent the actual information that's going into the
log is probably quite important.

I'd (for example) to be able to post-process a JSONEncoder log output
back into a readable format as if it had been printed by TextEncoder directly,
but couldn't do that if they print different things.

As for multi-line output, it would be good experiment with a few different formats
on some real logs to see how they look. It may end up looking nice, or perhaps
the extra lines will make for a log that harder to read because of all the extra
vertical space used.

**Comment by tesujimath** [comment]
I'd like to vote against multi-line output, at least unless there's always a single-line output available.  Grepping through a logfile is incredibly useful, and hopeless for multi-line output.

Also, TextEncoder is nice for human legible output as it is.  I suggest that any attempt to make it reversible is missing the point, and you should be using JSON encoder.

**Comment by akshayjshah** [comment]
Addressed in #267.

**Comment by akshayjshah** [comment]
Landed on dev.
